### ISSUE TITLE
Warnings-Turned Error in pathkeys.c under GCC 6+ `-O1`

### ISSUE THREAD

**Issue by d** [greenplum-db/gpdb#4123]
### Greenplum version or build
master (454c0e87b8)

### OS version and uname -a
Ubuntu 17.04 Zesty (kernel Linux 4.4)

### autoconf options used ( config.status --config )
./configure --enable-orca --with-gssapi --enable-mapreduce --with-perl --with-libxml --with-python --disable-gpcloud --disable-pxf --enable-gpfdist --prefix=/build/install --with-includes=/build/install/include --with-libs=/build/install/lib --enable-debug --enable-cassert 'CXX=ccache c++' 'CC=ccache cc' 'CFLAGS=-O1 -fno-omit-frame-pointer'

### Installation information ( pg_config )
N/A: can't compile Greenplum

### Expected behavior
It compiles ...

### Actual behavior
```
pathkeys.c: In function 'cdb_make_pathkey_for_expr':
pathkeys.c:1137:6: error: 'strategy' may be used uninitialized in this function [-Werror=maybe-uninitialized]
   pk = make_canonical_pathkey(root, eclass, opfamily, strategy, false);
   ~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: some warnings being treated as errors
make[4]: *** [pathkeys.o] Error 1
make[4]: Entering directory '/build/gpdb/src/backend/optimizer/path'
ccache cc -O1 -fno-omit-frame-pointer -std=gnu99  -Wall -Wmissing-prototypes -Wpointer-arith -Wendif-labels -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -fno-aggressive-loop-optimizations -Wno-unused-but-set-variable -Wno-address -g -ggdb  -Werror=uninitialized -I../../../../src/include -D_GNU_SOURCE -I/usr/include/libxml2  -I/build/install/include  -c -o pathkeys.o pathkeys.c
<builtin>: recipe for target 'pathkeys.o' failed
make[4]: Leaving directory '/build/gpdb/src/backend/optimizer/path'
../../../src/backend/common.mk:41: recipe for target 'path-recursive' failed
make[3]: *** [path-recursive] Error 2
common.mk:41: recipe for target 'optimizer-recursive' failed
```

### Step to reproduce the behavior
Use a compiler from GCC 6 Release Series (in this case 6.3 from Ubuntu 17.04), run `configure` with `--enable-cassert --enable-debug CFLAGS='-O1 -fno-omit-frame-pointer'`, then compile.

**Comment by d** [comment]
I was able to see the same warning-turned-error under GCC 7 as well, both on macOS and Ubuntu

**Comment by d** [comment]
@hlinnaka @danielgustafsson this seems to be introduced during the 8.3 merge in 087bf2e8919d4403c345846245ed0682e134416c, although a big part of it seems to be existing code (4.3) moved around. I can't tell whether the compiler is right that we could end up never assigning to `strategy` in function `cdb_make_pathkey_for_expr`

**Comment by danielgustafsson** [comment]
> On 12 Dec 2017, at 23:14, Jesse Zhang <notifications@github.com> wrote:
> 
> @hlinnaka @danielgustafsson this seems to be introduced during the 8.3 merge in 087bf2e, although a big part of it seems to be existing code (4.3) moved around. I can't tell whether the compiler is right that we could end up never assigning to strategy in function cdb_make_pathkey_for_expr

From a quick read, if the op_mergejoinable() guard returns true then get_mergejoin_opfamilies() will always return and strategy will be set. That being said, if it fails compilation it still needs additional work. Initializing strategy to zero to keep the compiler happy seems the least intrusive fix (even if I’m not a fan of such initializations).

**Comment by d** [comment]
Daniel, If I understand it correctly, I thought `op_mergejoinable` only looks at `pg_operator.oprcanmerge`, which is necessary but not sufficient condition for the access method strategy to exist: `get_mergejoin_opfamilies` searches in `pg_amop` by `amopopr`

**Comment by danielgustafsson** [comment]
Agreed, I wasn't paying enough attention when skimming last night. We should initialize the variable and catch any failure to find opfamilies along the lines of:
```diff
--- a/src/backend/optimizer/path/pathkeys.c
+++ b/src/backend/optimizer/path/pathkeys.c
@@ -1123,6 +1123,10 @@ cdb_make_pathkey_for_expr(PlannerInfo *root,
                elog(ERROR, "could not find mergejoinable = operator for type %u", typeoid);

        mergeopfamilies = get_mergejoin_opfamilies(eqopoid);
+       if (!mergeopfamilies)                   /* certainly should find some */
+               elog(ERROR, "could not find opfamilies for equality operator %u",
+                        ordering_op);
+
        foreach(lc, mergeopfamilies)
        {
                opfamily = lfirst_oid(lc);
```

**Comment by hlinnaka** [comment]
Fixed in commit 5b851540ab27e1321e6ea38282a91c4431349181
